llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp

```diff
@@ -2485,6 +2490,197 @@ bool ObjCARCOpt::run(Function &F, AAResults &AA) {
+          break;
+
+        auto &TopPool = PoolStack.back();
+        CallInst *PendingPush = TopPool.first;
+        bool HasAutoreleaseInScope = TopPool.second;
+
+        // Pop the stack - remove this pool scope
+        PoolStack.pop_back();
+
+        // Check if this pop matches the pending push by comparing the token
+        Value *PopArg = Pop->getArgOperand(0);
+        bool IsMatchingPop = (PopArg == PendingPush);
+
+        // Also handle pointer casts by stripping them
+        if (!IsMatchingPop) {
+          Value *StrippedPopArg = PopArg->stripPointerCasts();
+          IsMatchingPop = (StrippedPopArg == PendingPush);
+        }
+
+        if (!IsMatchingPop)
+          break;
+
+        // Before adding to EmptyPoolPairs, verify ALL uses of the push
+        // eventually lead to this specific pop instruction
+        bool SafeToOptimize = true;
+        for (User *U : PendingPush->users()) {
+          // Check if this use eventually leads to our pop instruction
+          bool LeadsToExpectedPop = false;
+
+          if (U == Pop)
+            LeadsToExpectedPop = true;
+          else if (isa<CastInst>(U)) {
+            // For pointer casts, check if they eventually lead to the pop
+            // (up to 5 levels deep)
+            SmallPtrSet<Value *, 8> Visited;
+            if (checkLeadsToTarget(cast<Value>(U), Pop, 5, Visited)) {
+              LeadsToExpectedPop = true;
+            }
+          }
+
+          if (!LeadsToExpectedPop) {
+            SafeToOptimize = false;
+            LLVM_DEBUG(dbgs() << "Unsafe to optimize: push has unexpected use: "
+                              << *U << "\n");
+            break;
+          }
+        }
+
+        if (!HasAutoreleaseInScope && SafeToOptimize) {
+          LLVM_DEBUG(dbgs() << "Eliminating empty autorelease pool pair: "
+                            << *PendingPush << " and " << *Pop << "\n");
+
+          // Store the pair for careful deletion later
+          EmptyPoolPairs.push_back({PendingPush, Pop});
+
+          ++NumNoops;
+        }
+        break;
+      }
+      case ARCInstKind::CallOrUser:
+      case ARCInstKind::Call:
+      case ARCInstKind::Autorelease:
+      case ARCInstKind::AutoreleaseRV: {
+        // Track that we have autorelease calls in the current pool scope
+        if (!PoolStack.empty()) {
+          PoolStack.back().second = true; // Set has_autorelease_in_scope = true
+          LLVM_DEBUG(
+              dbgs()
+              << "Found autorelease or potiential autorelease in pool scope: "
+              << Inst << "\n");
+        }
+        break;
+      }
+
+      // Enumerate all remaining ARCInstKind cases explicitly
+      case ARCInstKind::Retain:
+      case ARCInstKind::RetainRV:
+      case ARCInstKind::UnsafeClaimRV:
+      case ARCInstKind::RetainBlock:
+      case ARCInstKind::Release:
+      case ARCInstKind::NoopCast:
+      case ARCInstKind::FusedRetainAutorelease:
+      case ARCInstKind::FusedRetainAutoreleaseRV:
```

Comment:
`FusedRetainAutorelease` and `LoadWeak` can both autorelease. So can `FusedRetainAutoreleaseRV`, although really we should just never see that within an autorelease scope.

Reply:
Yeah but they aren't checked in other places regarding autorelease. are those bugged too? like isaautorelease does not return true on those.